business/workloads.go

```diff
@@ -1852,3 +1870,191 @@ func (in *WorkloadService) GetWorkloadAppName(ctx context.Context, namespace, wo
 	app := wkd.Labels[appLabelName]
 	return app, nil
 }
+
+// streamParsedLogs only works correctly when opts does not have Duration set || does not have TailLines set.
+func (in *WorkloadService) streamParsedLogs(namespace, name string, opts *LogOptions, w http.ResponseWriter) error {
+	k8sOpts := opts.PodLogOptions
+	// the k8s API does not support "endTime/beforeTime". So for bounded time ranges we need to
+	// discard the logs after sinceTime+duration
+	isBounded := opts.Duration != nil
+
+	logsReader, err := in.k8s.StreamPodLogs(namespace, name, &k8sOpts)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		e := logsReader.Close()
+		if e != nil {
+			log.Errorf("Error when closing the connection streaming logs of a pod: %s", e.Error())
+		}
+	}()
+
+	bufferedReader := bufio.NewReader(logsReader)
+
+	var startTime *time.Time
+	var endTime *time.Time
+	if k8sOpts.SinceTime != nil {
+		startTime = &k8sOpts.SinceTime.Time
+		if isBounded {
+			end := startTime.Add(*opts.Duration)
+			endTime = &end
+		}
+	}
+
+	engardeParser := parser.New(parser.IstioProxyAccessLogsPattern)
+
+	// To avoid high memory usage, the JSON will be written
+	// to the HTTP Response as it's received from the cluster API.
+	// That is, each log line is parsed, decorated with Kiali's metadata,
+	// marshalled to JSON and immediately written to the HTTP Response.
+	// This means that it is needed to push HTTP headers and start writing
+	// the response body right now and any errors at the middle of the log
+	// processing can no longer be informed to the client. So, starting
+	// these lines, the best we can do if some error happens is to simply
+	// log the error and strop/truncate the response, which will have the
+	// effect of sending an incomplete JSON document that the browser will fail
+	// to parse. Hopefully, the client/UI can catch the parsing error and
+	// properly show an error message about that there was a failure retrieving logs.
+	w.Header().Set("Content-Type", "application/json")
+	_, writeErr := w.Write([]byte("{\"entries\":[")) // This starts the JSON document
+	if writeErr != nil {
+		return writeErr
+	}
+
+	firstEntry := true
+	line, readErr := bufferedReader.ReadString('\n')
+	for ; readErr == nil || (readErr == io.EOF && len(line) > 0); line, readErr = bufferedReader.ReadString('\n') {
+		entry := LogEntry{
```

Comment:
I think probably a lot of the duplicated code for processing the individual log entries could be consolidated into a single function, leaving `streamParsedLogs` and `getParsedLogs` to more clearly focus on the different retrieval approaches.

Reply:
I agree. I'll check how to de-duplicate the code.

Reply:
OK, I extracted some code into a function.
I think it is still some room for code de-duplication, but that would require more of a refactoring.

I hope it is better.